src/test/java/com/google/devtools/build/lib/skylark/SkylarkDefinedAspectsTest.java

```diff
@@ -906,19 +906,13 @@ public void aspectFailingOrphanArtifacts() throws Exception {
 
   @Test
   public void aspectSkippingOrphanArtifactsWithLocation() throws Exception {
-    scratch.file("aspects/BUILD", "");
-    scratch.file(
-        "aspects/print.bzl",
-        "def _print_expanded_copts_impl(target, ctx):",
-        "    # Make sure the rule has a srcs attribute.",
-        "    if hasattr(ctx.rule.attr, 'cmd'):",
-        "        cmd = ctx.rule.attr.cmd",
-        "        print(\"{} -> {}\".format(cmd, ctx.expand_location(cmd, [])))",
-        "    return []",
+    scratch.file(
+        "simple/print.bzl",
+        "def _print_expanded_location_impl(target, ctx):",
+        "    return struct(result=ctx.expand_location(ctx.rule.attr.cmd, []))",
```

Comment:
Would this tickle the bug? I thought that the bug was that the *outputs* of the  rule in question are not expanded, and `$(location :files)` expands to the *inputs* of `:concat_all_files`. What would tickle the bug is `cmd = "$(location :concatenated.txt)"`.

Reply:
Yeah, it will tickle the bug. The bug is that expanding the _inputs_ of a dependent rule results in a build error. With Bazel 0.23.1, I put this test in a dir: 

```
 $ for i in simple/* ; do echo; echo === $i === ; cat $i ; done

=== simple/BUILD ===
filegroup(
    name = "files",
    srcs = ["afile"],
)

genrule(
    name = "concat_all_files",
    srcs = [":files"],
    outs = ["concatenated.txt"],
    cmd = "cat $(location :files) > $@"
)
=== simple/afile ===
some content
=== simple/print.bzl ===
def _print_expanded_copts_impl(target, ctx):
    cmd = ctx.rule.attr.cmd
    return struct(result=ctx.expand_location(ctx.rule.attr.cmd, []))

print_expanded_copts = aspect(
    implementation = _print_expanded_copts_impl,
)
```
and running it:

```
$ bazel build //simple:concat_all_files --aspects //simple:print.bzl%print_expanded_copts
INFO: Using Bazel Remote listening on 127.0.0.1:55064
INFO: Running Bazel version 0.23.1
INFO: Invocation ID: 745c9afc-31e7-4bef-94aa-da8c3b049aa0
ERROR: /Users/obonilla/bazel-bugs/7083/repro/simple/BUILD:6:1: in //simple:print.bzl%print_expanded_copts aspect on genrule rule //simple:concat_all_files: 


The following files have no generating action:
simple/concatenated.txt
ERROR: Analysis of aspect '//simple:print.bzl%print_expanded_copts of //simple:concat_all_files' failed; build aborted: Analysis of target '//simple:concat_all_files' failed; build aborted
INFO: Elapsed time: 0.086s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
```

Building it with the fix in this PR:

```
$ ~/o/bazel/bazel-bin/src/bazel build //simple:concat_all_files --aspects //simple:print.bzl%print_expanded_copts
Starting local Bazel server and connecting to it...
INFO: Analysed target //simple:concat_all_files (3 packages loaded, 20 targets configured).
INFO: Found 1 target...
Aspect //simple:print.bzl%print_expanded_copts of //simple:concat_all_files up-to-date (nothing to build)
INFO: Elapsed time: 2.279s, Critical Path: 0.01s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action```

Reply:
Actually, the truth is in the middle :) The bug happens when the aspect tries to expand the outputs of the rule it's attached to and I suppose when you say "dependent rule" you mean the rule the aspect depends on.

The test case does indeed tickle the bug but not because the erroneous label gets expanded but because the erroneous label gets into the list of labels that can be possibly expanded.

Reply:
Yeah, that's my understanding as well. The bug is that the aspect was expanding the outputs of the rule it's attached to unnecessarily (for this case). It's kind of subtle... thanks for all your help and patience.

Reply:
No worries! Thanks to y'all for fixing this bug.